File: java/src/test/java/com/orange/dojo/elephantcarpaccio/terminal/NumberReaderTest.java
package com.orange.dojo.elephantcarpaccio.terminal;

import java.io.StringReader;
import java.util.InputMismatchException;
import java.util.Scanner;

import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class NumberReaderTest {

  @Test
  public void I_Read_An_Integer() {
    // Given
    int expectedNumber = 14;

    Scanner input = new Scanner(new StringReader(String.format("%d", expectedNumber)));
    NumberReader numberReader = new NumberReader(input);

    // When
    int integer = numberReader.readInteger();

    // Then
    assertThat(integer).isEqualTo(expectedNumber);
  }

  @Test
  public void I_Read_A_Float() {
    // Given
    float expectedNumber = 14.4f;

    Scanner input = new Scanner(new StringReader(String.format("%f", expectedNumber)));
    NumberReader numberReader = new NumberReader(input);

    // When
    float aFloat = numberReader.readFloat();

    // Then
    assertThat(aFloat).isEqualTo(expectedNumber);
  }

  @Test(expected = InputMismatchException.class)
  public void when_The_Input_Is_Not_A_Number_It_Raises_An_InvalidInputException_if_we_expect_an_integer() {
    // Given
    Scanner input = new Scanner(new StringReader("f45"));
    NumberReader numberReader = new NumberReader(input);

    // When
    numberReader.readInteger();

    // Then

  }

  @Test(expected = InputMismatchException.class)
  public void when_The_Input_Is_Not_A_Number_It_Raises_An_InvalidInputException_if_we_expect_a_float() {
    // Given
    Scanner input = new Scanner(new StringReader("f45"));
    NumberReader numberReader = new NumberReader(input);

    // When
    numberReader.readFloat();

    // Then

  }

  @Test(expected = InputMismatchException.class)
  public void when_The_Input_Is_A_Float_It_Raises_An_InvalidInputException_if_we_expect_an_integer() {
    // Given
    Scanner input = new Scanner(String.format("%f", 24.5f));
    NumberReader numberReader = new NumberReader(input);

    // When
    int integer = numberReader.readInteger();

    // Then

  }
}

File: kotlin/src/test/kotlin/com/orange/dojo/elephantcarpaccio/terminal/NumberReaderTest.kt
package com.orange.dojo.elephantcarpaccio.terminal

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.io.StringReader
import java.util.*


internal class NumberReaderTest {

    @Test
    fun `I read an Integer`() {
        // Given
        val expectedNumber = 14

        val input = Scanner(StringReader(String.format("%d", expectedNumber)))
        val numberReader = NumberReader(input)

        // When
        val integer = numberReader.readInteger()

        // Then
        assertThat(integer).isEqualTo(expectedNumber)
    }

    @Test
    fun `I read a Float`() {
        // Given
        val expectedNumber = 14.4f

        val input = Scanner(StringReader(String.format("%f", expectedNumber)))
        val numberReader = NumberReader(input)

        // When
        val aFloat = numberReader.readFloat()

        // Then
        assertThat(aFloat).isEqualTo(expectedNumber)
    }


    @Test
    fun `When the input is not a number it raises an InputMismatchException if we expect an Integer`() {
        // Given
        val input = Scanner(StringReader("f45"))
        val numberReader = NumberReader(input)

        // Then
        assertThrows<InputMismatchException> {
            // When
            numberReader.readInteger()
        }
    }

    @Test
    fun `When the input is not a number it raises an InputMismatchException if we expect a Float`() {
        // Given
        val input = Scanner(StringReader("f45"))
        val numberReader = NumberReader(input)

        // Then
        assertThrows<InputMismatchException> {
            // When
            numberReader.readFloat()
        }
    }

    @Test
    fun `When the input is a Float it raises an InputMismatchException if we expect an Integer`() {
        // Given
        val input = Scanner(String.format("%f", 24.5f))
        val numberReader = NumberReader(input)

        // Then
        assertThrows<InputMismatchException> {
            // When
            numberReader.readInteger()
        }
    }
}
